math/vc_soa/include/algebra/math/impl/vc_soa_transform3.hpp

```diff
@@ -0,0 +1,386 @@
+/** Algebra plugins library, part of the ACTS project
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "algebra/math/cmath.hpp"
+#include "algebra/qualifiers.hpp"
+#include "algebra/storage/vector.hpp"
+
+// Vc include(s).
+#ifdef _MSC_VER
+#pragma warning(push, 0)
+#endif  // MSVC
+#include <Vc/Vc>
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif  // MSVC
+
+// System include(s).
+#include <cassert>
+#include <type_traits>
+
+namespace algebra::vc_soa::math {
+
+using algebra::storage::operator*;
+using algebra::storage::operator/;
+using algebra::storage::operator-;
+using algebra::storage::operator+;
+
+namespace internal {
+
+/// 4x4 matrix type used by @c algebra::vc_soa::math::transform3 that has simd
+/// vectors as matrix elements
+template <typename scalar_t, template <typename, std::size_t> class array_t>
+struct matrix44 {
```

Comment:
Why don't you use the stroage matrix type of `vc_soa` algebra?

Reply:
I am ```using vector_type = storage::vector<3, Vc::Vector<scalar_t>, array_t>;``` for the ```matrix44``` columns. This way I can use e.g. the cross product on them and easily build a matrix from ```vector3``` types. If I use ```storage::vector<ROWS * COLS>```, I just have a ```Vc::Vector<T>``` type for every matrix element, for which I have no linear algebra implementation. 
So say I want to implement ```rotate```: Right now that is simple but with ```storage::matrix``` I would have to write: {m[0][0] * v[0] + m[0][1] * v[1] + m[0][2] * v[2], m[1][0] * v[0] + m[1][1] * v[1] + m[1][2] * v[2], ...}```

Reply:
I suppose It is because there is no matrix operation for `vc_soa::matrix_type`
Is it possible to make one for `vc_soa::matrix_type` or it will just degrade the performance of SoA?

Reply:
Mainly, a matrix implementation would be a lot of extra work at this point which is not needed so far. If there is a usecase in traccc we can add it later

The last row of that matrix would eat a bit of memory: E.g. on my laptop [[0,0,0,0,0,0,0,0], [0,0,0,0,0,0,0,0], [0,0,0,0,0,0,0,0], [1,1,1,1,1,1,1,1]] for both the matrix and the inverse matrix per surface. But we could make it contiguous in memory in a rowwise layout, so it should not be that bad in terms of memory access (it could be that this block will then simply not be loaded since the transform3 implementation does not access it). 

However, I don't think it is possible for the AoS Vc implementation to do that (we might be able to share ```matrix44``` between the two plugins later, but I am not sure yet)

Reply:
I guess `vc_soa` is being implemented for vectorization, isn't it? If so, it should support matrix operation as well as vector operation (Matrix operation is a thing that you might want to optimize the most...)

> The last row of that matrix would eat a bit of memory: E.g. on my laptop [[0,0,0,0,0,0,0,0], [0,0,0,0,0,0,0,0], [0,0,0,0,0,0,0,0], [1,1,1,1,1,1,1,1]] 

Is it from `matrix44` or Is it a thing that going to happen with `vc_soa::matrix_type`?

> I don't think it is possible for the AoS Vc implementation

Is `AoS Vc implementation` vc_vc frontend we have? We can forget about this frontend as it is not what we want from vc plugin

Reply:
That would happen with ```vc_soa::matrix_type``` if we don't make it an asymmetric matrix (which would be OK, I guess, since in SoA the memory alignment is independent from the matrix dimension). 
Yes, it is implemented for vectorization, but in detray at least I did not find a good spot for SoA matrices, yet? Even putting the static geometry into SoA layout is not easy...
I would rather go with an AoS vectorization scheme (```vc``` backend) for the propagation in detray (at least until we come up with a good idea so handle the track state divergence), so next step would be fixing the ```vc``` plugin and then adding a matrix implemenation to it. Why is the ```vc_vc``` frontend not what we want?

Reply:
> in detray at least I did not find a good spot for SoA matrices, yet?

Using SoA matrix for covariance transport would be as interesting as usiing SoA vector for intersector

> Why is the vc_vc frontend not what we want?

I don't know its future form, but it is not using any actual `vc` function currently. Can it be more interesting in the future? Vc can be efficient for both AoS and SoA? My understanding on Vc plugin is not great...

Reply:
Yes, AoS and SoA are just memory layouts, while Vc is a library for explicit vectorization. Vectorization usually works better in an SoA layout, because you can fill the vector registers with more relevant data, but the old Vc plugin can vectorize the data in AoS layout and also achieve some speedup (especially 8x8 matrices with avx2).

The old Vc plugin implementation is fine actually, it was just incorrectly adapted to the cmath matrix actor, which broke it. But if we remove that matrix actor and write one that uses ```Vc::SimdArray``` internally, then we can use it for AoS vectorization. However, as far as I know, ```Vc::SimdArray``` is not made for linear algebra vectors that are much larger than the underlying vector register capacity, so for a generalized N-dimensional implementation, this could become a bit complicated.

Due to the track state divergence, I don't think an SoA memory layout makes sense to use in the covariance transport until we find a way to implement it such that we can pretty much guarantee that we can sort the tracks into chunks of 8, 16 , 32 etc. after seeding and then always access these tracks together in memory and have them follow the same algorithmic code path as well (or did you find a way in the CKF maybe?). Or alternatively, do you know of a way to use intra-track parallelization with SoA in the covariance transport (I thought the covariance matrices of a single track are strictly dependent on each other)?

Reply:
>  (or did you find a way in the CKF maybe?).

Is that correct that you are using vc_soa to calculate the intersections of multiple candidates efficiently during navigator initialization? 
(Actually I don't understand what is the  difference between original intersector and SoA intersector you introduced in acts-project/detray#555)

If so, we can also do the KF for multiple measurements during CKF. I am not sure if you have looked into the results of [mkFit](https://iris-hep.org/projects/mkfit.html) of CMS collaboration, but track building part is the only place they achieved the speedup with vectorized matrix.

Reply:
> Is that correct that you are using vc_soa to calculate the intersections of multiple candidates efficiently

Yes, that is the plan, but the navigator has not been adapted to use these intersectors, yet. The CPU will be capable of intersecting multiple surfaces for a single track in parallel. But which surfaces can be intersected in parallel is pretty much a fixed thing.

> If so, we can also do the KF for multiple measurements during CKF. I am not sure if you have looked into the results of [mkFit](https://iris-hep.org/projects/mkfit.html) of CMS collaboration, but track building part is the only place they achieved the speedup with vectorized matrix.

Cool, I did not know that! That is definetly something we should consider then, can we go over their paper on Monday?

Reply:
yes we can I guess. Andi must know about mkFit well